Explain stack overflow for nat, number examples (#17406)
coq/coq: doc/sphinx/language/core/basic.rst

     hexnat ::= {| 0x | 0X } @hexdigit {* {| @hexdigit | _ } }
     hexdigit ::= {| 0 .. 9 | a .. f | A .. F }

  .. todo PR: Need an example of when the following paragraph matters

  :n:`@integer` and :n:`@natural` are limited to the range that fits
  into an OCaml integer (63-bit integers on most architectures).

>   :n:`@bigint` and :n:`@bignat` have no range limitation.

Author: FWIW, I see the following doesn't play nicely with `Fail`.  The range check is probably in the parser before any `Fail` semantics are recognized.

![image](https://user-images.githubusercontent.com/1253341/226416239-1c561da4-fc95-4724-a796-6b2280b031d6.png)

>   :n:`@bigint` and :n:`@bignat` have no range limitation.

Reviewer: Well, I guess it's fine, it's not really a failure of the command but of Coq parser itself.

     hexnat ::= {| 0x | 0X } @hexdigit {* {| @hexdigit | _ } }
     hexdigit ::= {| 0 .. 9 | a .. f | A .. F }

>   :n:`@bigint` and :n:`@bignat`, which are used in :token:`term`\s, generally

Reviewer: ```suggestion
  :n:`@number`, :n:`@bigint` and :n:`@bignat`, which are used in :token:`term`\s, generally
```

  have no range limitation.

>   :n:`@integer` and :n:`@natural`, which are used as arguments in some tactics

Reviewer: maybe
```suggestion
  :n:`@integer` and :n:`@natural`, which are used as arguments in tactics
```
as it is rather "almost all tactics"

  The :cmd:`Number Notation` mechanism offers the user
  a way to define custom parsers and printers for :n:`@number`.

  By default, numbers are interpreted as :n:`nat`\s, which is a unary
  representation.  For example, :n:`3` is represented as `S (S (S O))`.  While
  this is a convenient representation for doing proofs, computing with large
  :n:`nat`\s can lead to stack overflows or running out of memory.  You can
  explicitly specify a different interpretation to avoid this problem.  For

>   example, :n:`1000000%N` efficiently represents that natural number.  See

Reviewer: ```suggestion
  example, :n:`1000000%N` is a more efficient binary representation of that natural number.  See
```
But maybe rather take the more common Z as an example and link to https://coq.gitlab.io/-/coq/-/jobs/3977290672/artifacts/_build/default/doc/refman-html/language/coq-library.html#notations-for-integer-arithmetic

Author: The hyperlink points to info about notations for operations rather than for numbers, so I think not helpful here.  Too much detail.

  :ref:`Scopes` and :n:`@term_scope`.  You can also change the default

>   interpretation using the :cmd:`Number Notation` command.

Reviewer: This sentence is already at the end of the previous paragraph. Moreover, the NUmber Notation command rather targets a few specific users (library authors (honestly, even me as coauthor of the command, I probably don't use it more than once a year ;) )). The casual user is probably more interested in the fact that multiple libraries offer her a wide range of options. Maybe something along the lines: "Multiple libraries offer numbers with different ease of proof/efficiency of computation trade-offs, ranging from the unary nat in the stdlib (https://coq.gitlab.io/-/coq/-/jobs/3977290672/artifacts/_build/default/doc/refman-html/language/coq-library.html#basic-arithmetic) to arbitrary precision machine integers in the bignums library (https://github.com/coq/bignums/)".

Author: Yeah, doesn't need to be repeated, I will adjust.

How does `Z` differ to the user from bignums?

I'm reluctant to start enumerating alternative libraries here.  There should be another section elsewhere that mentions all the alternatives and how to choose between them, which we could link to if it existed.  This part of the chapter primarily introduces syntax.  Getting into specific libraries is too much detail.  I don't want to expand the scope of the PR to create such a section right now.

Reviewer: > How does `Z` differ to the user from bignums?

`Z` are binary integers, that is one constructor (a few machine words) for each bit of the number where bignums offers a machine integer encoding, that is a constructor for each packet of 63 bits. The second is then much more efficient computationally speaking.

> I'm reluctant to start enumerating alternative libraries here. There should be another section elsewhere that mentions all the alternatives and how to choose between them, which we could link to if it existed. This part of the chapter primarily introduces syntax. Getting into specific libraries is too much detail. I don't want to expand the scope of the PR to create such a section right now.

I agree a full list would require work, I was just expecting we could add a few, definitely non exhaustive, examples to make clear to the reader that numbers are a library thing and that multiple options are available around (which is probably something not that common among usual programming languages, hence the need to pinpoint it IMHO).

>    .. example:: Stack overflow with :n:`nat`

Reviewer: The example is great but it offers more information than the authoritative source (which in this case I think is https://coq.gitlab.io/-/coq/-/jobs/3977290672/artifacts/_build/default/doc/refman-html/language/coq-library.html#basic-arithmetic this seem to indicate that the authoritative source should also be updated (and linked to)

Author: >  the authoritative source should also be updated (and linked to)

Absolutely, but I don't want to take that on right now.  We need a section that introduces number concepts and mentions the most important alternatives, synthesizing the mass of detail (which I think also needs much work.

Reviewer: Sorry if I made it sound like I was expecting you to fully rewrite the stdlib doc. I was just expecting it would be easy to add that small piece of information there. But maybe you are right and it's not that trivial after all.

      .. coqtop:: all reset

>          Fail Compute 100000 + 100000.  (* gives a stack overflow (not shown) *)

Reviewer: ```suggestion
         Fail Eval compute in 100000 + 100000.  (* gives a stack overflow (not shown) *)
```
(equivalent but `Compute` probably only remains for historical reasons from an old time there was only one evaluation mechanism in Coq)

Author: I think `Compute` is widely used and it's simpler.  Software Foundations, for example, uses `Compute`.

      .. coqtop:: in

         Require Import NArith.  (* for definition of N *)

      .. coqtop:: all

         Compute (1 + 1000000000000000000000000000000000)%N.

> 

Reviewer: This is about the stdlib rather than Coq itself so it should not be put/duplicated here. Maybe a short sentence like: "Coq itself is agnostic about numbers. Various libraries, including the default-loaded prelude (link to https://coq.inria.fr/distrib/current/refman/language/coq-library.html#basic-arithmetic ), provide various number interpretations."

Any further documentation about the stdlib should go into its documentation (for instance https://coq.inria.fr/distrib/current/refman/language/coq-library.html#basic-arithmetic )
